contracts/AddressProviderNG.vy

```diff
@@ -0,0 +1,290 @@
+    _tags: DynArray[String[64], 20],
+    _description: String[64],
+):
+    """
+    @notice Enter a new registry item
+    @param _address Address assigned to the _id
+    @param _id Address assigned to the input _id
+    @param _tags tags defining the entry. e.g. ['StableSwap', 'Factory']
+                 Entry can have a maximum of 20 tags. Tags need to be
+                 added to the AddressProvider via `add_new_tags` before id can be added.
+    @param _description Human-readable description of the identifier
+    """
+    assert msg.sender == self.admin  # dev: admin-only function
+    assert not self.check_id_exists[_id]  # dev: id exists
+    assert len(_tags) > 0  # dev: entry needs at least one tag
+
+    self.check_id_exists[_id] = True
+    self.ids.append(_id)
+
+    # Check if tags are correct and add them to tag > id mapping:
+    for _tag in _tags:
+        assert self.check_tag_exists[_tag]  # dev: unrecognised tag
+        self.id_tag_mapping[
+            keccak256(concat(uint2str(_id), _tag))
+        ] = True
+
+    # Add entry:
+    self.get_id_info[_id] = AddressInfo(
+        {
+            addr: _address,
+            description: _description,
+            tags: _tags,
+            version: 1,
+            last_modified: block.timestamp,
+        }
+    )
+    self.num_entries += 1
+
+    log NewEntry(_id, _address, _description)
+
+
+@external
+def update_address(_id: uint256, _address: address):
+    """
+    @notice Set a new address for an existing identifier
+    @param _id Identifier to set the new address for
+    @param _address Address to set
+    """
+    assert msg.sender == self.admin  # dev: admin-only function
+
+    # Update address:
+    self.get_id_info[_id].addr = _address
+
+    # Update metadata (version, update time):
+    self._update_entry_metadata(_id)
+
+
+@external
+def update_description(_id: uint256, _description: String[256]):
+    """
+    @notice Update description for an existing _id
+    @param _id Identifier to set the new description for
+    @param _description New description to set
+    """
+    assert msg.sender == self.admin  # dev: admin-only function
+    assert self.get_id_info[_id].addr != empty(address)  # dev: id is empty
+
+    # Update description:
+    self.get_id_info[_id].description = _description
+
+    # Update metadata (version, update time):
+    self._update_entry_metadata(_id)
+
+
+@external
+def update_tags(_id: uint256, _tags: DynArray[String[64], 20]):
+    """
+    @notice Update tags for an existing _id
+    @param _id Identifier to set the new description for
+    @param _tags New tags to set to
+    """
+    assert msg.sender == self.admin  # dev: admin-only function
+    assert self.get_id_info[_id].addr != empty(address)  # dev: id is empty
+
+    # Update tags:
+    self.get_id_info[_id].tags = _tags
+
+    # Update metadata (version, update time):
+    self._update_entry_metadata(_id)
+
+
+
+@external
+def remove_id(_id: uint256) -> bool:
```

Comment:
it would be good to have a corresponding `update_id` method that lets the admin update all fields at the same time. If an entry is updated, an integrator might expect or need all info to change in the same block. Instead they have to monitor the events for several blocks before they can get the rest of the info, with no guarantee as to what order they're coming in (are tags updated before or after the address?) A utility function that updates all fields at the same time seems more appropriate.